vyper/compiler/output_bundle.py

```diff
@@ -0,0 +1,232 @@
+    for i, s in enumerate(PurePath(p).parts):
+        if s == "..":
+            segments.append(str(i))
+        else:
+            segments.append(s)
+    return str(PurePath(*segments))
+
+
+# data structure containing things that should be in an output bundle,
+# which is some container containing the information required to
+# reproduce a build
+@dataclass
+class OutputBundle:
+    def __init__(self, compiler_data: CompilerData):
+        self.compiler_data = compiler_data
+
+    @cached_property
+    def compilation_target(self):
+        return self.compiler_data.compilation_target._metadata["type"]
+
+    @cached_property
+    def _imports(self):
+        return self.compilation_target.reachable_imports
+
+    @cached_property
+    def compiler_inputs(self) -> dict[str, CompilerInput]:
+        inputs: list[CompilerInput] = [
+            t.compiler_input for t in self._imports if not _is_builtin(t.qualified_module_name)
+        ]
+        inputs.append(self.compiler_data.file_input)
+
+        sources = {}
+        for c in inputs:
+            path = os.path.relpath(str(c.resolved_path))
+            # note: there should be a 1:1 correspondence between
+            # resolved_path and source_id, but for clarity use resolved_path
+            # since it corresponds more directly to zipfile semantics.
+            sources[_anonymize(path)] = c
+
+        return sources
+
+    @cached_property
+    def compilation_target_path(self):
+        p = self.compiler_data.file_input.resolved_path
+        p = os.path.relpath(str(p))
+        return _anonymize(p)
+
+    @cached_property
+    def used_search_paths(self) -> list[str]:
+        # report back which search paths were "actually used" in this
+        # compilation run. this is useful mainly for aesthetic purposes,
+        # because we don't need to see `/usr/lib/python` in the search path
+        # if it is not used.
+        # that being said, we are overly conservative. that is, we might
+        # put search paths which are not actually used in the output.
+
+        search_paths = self.compiler_data.input_bundle.search_paths
+
+        tmp = {sp: 0 for sp in search_paths}
+
+        for c in self.compiler_inputs.values():
+            # recover the search path that was used for this CompilerInput.
+            # note that it is not sufficient to thread the "search path that
+            # was used" into CompilerInput because search_paths are modified
+            # during compilation (so a search path which does not exist in
+            # the original search_paths set could be used for a given file).
+            for sp in reversed(search_paths):
+                if c.resolved_path.is_relative_to(sp):
+                    # don't break here. if there are more than 1 search path
+                    # which could possibly match, we add all them to the
+                    # output.
+                    tmp[sp] += 1
+
+        sps = [sp for sp, count in tmp.items() if count > 0]
```

Comment:
We only care if count>0, this could be a set
```suggestion
        sps = set()

        for c in self.compiler_inputs.values():
            # recover the search path that was used for this CompilerInput.
            # note that it is not sufficient to thread the "search path that
            # was used" into CompilerInput because search_paths are modified
            # during compilation (so a search path which does not exist in
            # the original search_paths set could be used for a given file).
            for sp in reversed(search_paths):
                if c.resolved_path.is_relative_to(sp):
                    # don't break here. if there are more than 1 search path
                    # which could possibly match, we add all them to the
                    # output.
                    sps.add(sp)
```

Or even in a single comprehension (sorry I know you don't like them but it's very succinct.
```suggestion
        sps = {
            sp
            for c in self.compiler_inputs.values():
            for sp in reversed(search_paths):
            if c.resolved_path.is_relative_to(sp):
        }
```

Reply:
these don't preserve the precedence of the search path.